text/3742-mutate-non-mutable-locals.md

```diff
@@ -0,0 +1,249 @@
+- Feature Name: `mut_non_mut`
+- Start Date: 2024-12-13
+- RFC PR: [rust-lang/rfcs#3742](https://github.com/rust-lang/rfcs/pull/3742)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+the `mut` keyword has two meanings currently:
+ - When applied to a reference, it means the reference is not aliased
+ - When applied to a local binding, it means that the binding is allowed be re-assigned or borrowed using `&mut`
+
+The first of these is fundamental to Rust's safety guarantees and enables some optimizations.  The second,
+however, is more of a lint: the error message may help catch mistakes, but ignoring it doesn't violate
+type safety or subvert any of Rust's global correctness guarantees.
+
+This RFC proposes that the re-assignment or creation of a `&mut` to a local variable not
+marked with `mut` be changed from a hard error to a deny-by-default lint diagnostic (called
+`mut_non_mut` for the sake of discussion).
+
+# Motivation
+[motivation]: #motivation
+
+This idea has been discussed previously:
+
+ - [2014 Blog post](https://smallcultfollowing.com/babysteps/blog/2014/05/13/focusing-on-ownership/)
+ - [2021 IRLO thread](https://internals.rust-lang.org/t/lack-of-mut-in-bindings-as-a-deny-by-default-lint/15818)
+ - [Zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/Idea.3A.20Downgrade.20.22immutable.20variable.E2.80.9C.20errors.20to.20lints)
+
+One benefit of this change is reduced friction: adding and removing of the need for `mut` on a given local
+can happen frequently when refactoring, similar to how variables and functions may change from used to unused
+and back.  In this context, a fatal error for `mut_non_mut` can be a flow interruption.  Yes, adding
+`mut` in all the right places is ideally something you should do _eventually_, just as you should
+delete any unused variables/functions, but it's not something that needs to prevent you from running
+the program.
```

Comment:
The "you should do this eventually" prescriptivism is unnecessary here. As @Lokathor [points out](https://github.com/rust-lang/rfcs/pull/3742#issuecomment-2544122660), a subset of Rust users (including myself) find this diagnostic to be pure annoyance, both bloating run-edit-cycles and adding line noise. Given that this RFC will allow such users to opt-out of this annoyance, there's no need to alienate this body of potential support by saying their preference is not ideal.

Reply:
I tweaked the phrasing a bit.  I'm still not super happy with the sentence, but it's definitely better without the "you should"s in it.